File: docs/beginner/module-02-basic-search/code-samples/csharp/05_FieldSearch.cs
/*
Field-Specific Search - Module 2 C# Examples
Searching within specific fields in Azure AI Search using .NET SDK

This module demonstrates:
- Searching specific fields
- Field selection for results
- Multi-field searches
- Field weighting concepts
- When to use field-specific search
*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Models;

namespace AzureSearchHandbook.Module02.BasicSearch
{
    public class FieldSearch
    {
        private readonly SearchClient _searchClient;

        public FieldSearch(SearchClient searchClient)
        {
            _searchClient = searchClient ?? throw new ArgumentNullException(nameof(searchClient));
        }

        /// <summary>
        /// Search within a specific field only
        /// </summary>
        /// <param name="query">Search query string</param>
        /// <param name="field">Field name to search in</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> SearchSpecificFieldAsync(string query, string field, int top = 10)
        {
            try
            {
                Console.WriteLine($"Searching field '{field}' for: '{query}'");

                var searchOptions = new SearchOptions
                {
                    SearchFields = { field },
                    Size = top,
                    IncludeTotalCount = true
                };

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results in field '{field}'");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error searching field '{field}': {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search within multiple specific fields
        /// </summary>
        /// <param name="query">Search query string</param>
        /// <param name="fields">List of field names to search in</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results</returns>
        public async Task<SearchResults<SearchDocument>> SearchMultipleFieldsAsync(string query, IEnumerable<string> fields, int top = 10)
        {
            try
            {
                var fieldList = fields.ToList();
                Console.WriteLine($"Searching fields [{string.Join(", ", fieldList)}] for: '{query}'");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                foreach (var field in fieldList)
                {
                    searchOptions.SearchFields.Add(field);
                }

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results in specified fields");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error searching multiple fields: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Search and return only selected fields
        /// </summary>
        /// <param name="query">Search query string</param>
        /// <param name="selectFields">List of field names to return in results</param>
        /// <param name="searchFields">List of field names to search in (optional)</param>
        /// <param name="top">Maximum number of results to return</param>
        /// <returns>Search results with only selected fields</returns>
        public async Task<SearchResults<SearchDocument>> SearchWithSelectedFieldsAsync(
            string query, IEnumerable<string> selectFields, IEnumerable<string> searchFields = null, int top = 10)
        {
            try
            {
                var selectList = selectFields.ToList();
                Console.WriteLine($"Searching for '{query}' with selected fields: [{string.Join(", ", selectList)}]");

                var searchOptions = new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                };

                foreach (var field in selectList)
                {
                    searchOptions.Select.Add(field);
                }

                if (searchFields != null)
                {
                    foreach (var field in searchFields)
                    {
                        searchOptions.SearchFields.Add(field);
                    }
                }

                var results = await _searchClient.SearchAsync<SearchDocument>(query, searchOptions);
                
                Console.WriteLine($"Found results with selected fields");
                return results.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in search with selected fields: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Compare search results across different fields
        /// </summary>
        /// <param name="query">Search query string</param>
        /// <param name="fields">List of fields to compare</param>
        /// <param name="top">Maximum results per field</param>
        /// <returns>Dictionary mapping field names to their results</returns>
        public async Task<Dictionary<string, SearchResults<SearchDocument>>> CompareFieldSearchesAsync(
            string query, IEnumerable<string> fields, int top = 5)
        {
            var results = new Dictionary<string, SearchResults<SearchDocument>>();

            // Search all fields (default)
            try
            {
                var allResults = await _searchClient.SearchAsync<SearchDocument>(query, new SearchOptions
                {
                    Size = top,
                    IncludeTotalCount = true
                });
                results["all_fields"] = allResults.Value;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error in all fields search: {ex.Message}");
                results["all_fields"] = null;
            }

            // Search individual fields
            foreach (var field in fields)
            {
                results[field] = await SearchSpecificFieldAsync(query, field, top);
            }

            return results;
        }

        /// <summary>
        /// Display comparison of field-specific searches
        /// </summary>
        /// <param name="query">Original search query</param>
        /// <param name="results">Results from different field searches</param>
        public static void DisplayFieldComparison(string query, 
            Dictionary<string, SearchResults<SearchDocument>> results)
        {
            Console.WriteLine($"\n🎯 Field-Specific Search Comparison: '{query}'");
            Console.WriteLine(new string('=', 70));

            foreach (var kvp in results)
            {
                var fieldName = kvp.Key;
                var fieldResults = kvp.Value;
                var resultList = fieldResults?.GetResults()?.ToList() ?? new List<SearchResult<SearchDocument>>();

                Console.WriteLine($"\n{fieldName.Replace("_", " ").ToTitleCase()}:");
                Console.WriteLine($"   Results found: {resultList.Count}");

                if (resultList.Any())
                {
                    Console.WriteLine("   Top matches:");
                    for (int i = 0; i < Math.Min(3, resultList.Count); i++)
                    {
                        var result = resultList[i];
                        var title = result.Document.TryGetValue("title", out var titleValue) ? 
                            titleValue?.ToString() : "No title";
                        var score = result.Score ?? 0.0;
                        Console.WriteLine($"     {i + 1}. {title} (Score: {score:F3})");
                    }

                    // Show average score
                    var avgScore = resultList.Where(r => r.Score.HasValue).Average(r => r.Score.Value);
                    Console.WriteLine($"   Average score: {avgScore:F3}");
                }
                else
                {
                    Console.WriteLine("   No matches found");
                }

                Console.WriteLine(new string('-', 50));
            }

            // Analysis
            Console.WriteLine($"\n📊 FIELD ANALYSIS:");
            foreach (var kvp in results)
            {
                var fieldName = kvp.Key;
                var fieldResults = kvp.Value;
                var count = fieldResults?.GetResults()?.Count() ?? 0;
                
                if (count > 0)
                {
                    var resultList = fieldResults.GetResults().ToList();
                    var maxScore = resultList.Where(r => r.Score.HasValue).Max(r => r.Score.Value);
                    Console.WriteLine($"   {fieldName}: {count} results (max score: {maxScore:F3})");
                }
                else
                {
                    Console.WriteLine($"   {fieldName}: No results");
                }
            }
        }
    }

    /// <summary>
    /// Extension method for title case conversion
    /// </summary>
    public static class StringExtensions
    {
        public static string ToTitleCase(this string input)
        {
            if (string.IsNullOrEmpty(input))
                return input;

            var words = input.Split(' ');
            for (int i = 0; i < words.Length; i++)
            {
                if (words[i].Length > 0)
                {
                    words[i] = char.ToUpper(words[i][0]) + words[i].Substring(1).ToLower();
                }
            }
            return string.Join(" ", words);
        }
    }

    /// <summary>
    /// Program class for demonstration
    /// </summary>
    public class FieldSearchProgram
    {
        // Replace with your actual service details
        private const string ServiceEndpoint = "https://your-service.search.windows.net";
        private const string ApiKey = "your-api-key";
        private const string IndexName = "your-index-name";

        public static async Task Main(string[] args)
        {
            try
            {
                // Initialize search client
                var searchClient = new SearchClient(
                    new Uri(ServiceEndpoint),
                    IndexName,
                    new AzureKeyCredential(ApiKey)
                );

                await DemonstrateFieldSearchAsync(searchClient);
                await FieldSearchBestPracticesAsync(searchClient);

                Console.WriteLine("\n💡 Next Steps:");
                Console.WriteLine("   - Try searching different fields in your index");
                Console.WriteLine("   - Experiment with field combinations");
                Console.WriteLine("   - Check out other examples for more search options");
                Console.WriteLine("   - Learn about result processing techniques");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"❌ Demo failed: {ex.Message}");
                Console.WriteLine("Make sure your Azure AI Search service is configured correctly.");
            }
        }

        private static async Task DemonstrateFieldSearchAsync(SearchClient searchClient)
        {
            Console.WriteLine("🎯 Field-Specific Search Demonstration");
            Console.WriteLine(new string('=', 50));

            var searchOps = new FieldSearch(searchClient);

            // Example 1: Compare field searches
            Console.WriteLine("\n1️⃣ Field Search Comparison");

            var query = "python";
            var commonFields = new[] { "title", "content", "description", "author" };

            var fieldResults = await searchOps.CompareFieldSearchesAsync(query, commonFields, 5);
            FieldSearch.DisplayFieldComparison(query, fieldResults);

            // Example 2: Multi-field search
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n2️⃣ Multi-Field Search Examples");
            Console.WriteLine(new string('-', 30));

            var multiFieldExamples = new[]
            {
                new { Description = "Find tutorials by title or description", Fields = new[] { "title", "description" } },
                new { Description = "Search content and tags", Fields = new[] { "content", "tags" } },
                new { Description = "Author and title search", Fields = new[] { "author", "title" } }
            };

            foreach (var example in multiFieldExamples)
            {
                Console.WriteLine($"\n📋 {example.Description}");
                Console.WriteLine($"   Fields: {string.Join(", ", example.Fields)}");

                var results = await searchOps.SearchMultipleFieldsAsync("tutorial", example.Fields, 3);
                var resultList = results.GetResults().ToList();
                Console.WriteLine($"   Results: {resultList.Count} found");

                if (resultList.Any())
                {
                    var topResult = resultList.First();
                    var title = topResult.Document.TryGetValue("title", out var titleValue) ? 
                        titleValue?.ToString() : "No title";
                    var score = topResult.Score ?? 0.0;
                    Console.WriteLine($"   Top match: {title} (Score: {score:F3})");
                }
            }

            // Example 3: Field selection
            Console.WriteLine($"\n{new string('=', 70)}");
            Console.WriteLine("\n3️⃣ Field Selection Examples");
            Console.WriteLine(new string('-', 30));

            var selectionExamples = new[]
            {
                new { Description = "Basic info only", SelectFields = new[] { "id", "title", "author" } },
                new { Description = "Content preview", SelectFields = new[] { "title", "description", "url" } },
                new { Description = "Metadata only", SelectFields = new[] { "id", "publishedDate", "category" } }
            };

            foreach (var example in selectionExamples)
            {
                Console.WriteLine($"\n📋 {example.Description}");
                Console.WriteLine($"   Selected fields: {string.Join(", ", example.SelectFields)}");

                var results = await searchOps.SearchWithSelectedFieldsAsync(
                    "programming", 
                    example.SelectFields, 
                    top: 2
                );

                var resultList = results.GetResults().ToList();
                Console.WriteLine($"   Results: {resultList.Count} found");

                if (resultList.Any())
                {
                    var firstResult = resultList.First();
                    var availableFields = firstResult.Document.Keys.Where(k => !k.StartsWith("@")).ToList();
                    Console.WriteLine($"   Available fields in result: {string.Join(", ", availableFields)}");
                }
            }

            Console.WriteLine("\n✅ Field-specific search demonstration completed!");
        }

        private static async Task FieldSearchBestPracticesAsync(SearchClient searchClient)
        {
            Console.WriteLine("\n📚 Field-Specific Search Best Practices");
            Console.WriteLine(new string('=', 50));

            Console.WriteLine("\n💡 When to Use Field-Specific Search:");
            Console.WriteLine("\n✅ Search Specific Fields When:");
            Console.WriteLine("   - You know exactly which field contains relevant information");
            Console.WriteLine("   - You want to search titles only (more precise)");
            Console.WriteLine("   - You need to search metadata fields (author, category, etc.)");
            Console.WriteLine("   - You want to exclude certain fields from search");

            Console.WriteLine("\n✅ Select Specific Fields When:");
            Console.WriteLine("   - You only need certain fields in results (faster)");
            Console.WriteLine("   - You want to reduce network traffic");
            Console.WriteLine("   - You're building a summary or preview");
            Console.WriteLine("   - You need consistent result structure");

            Console.WriteLine("\n⚠️ Field Search Considerations:");
            Console.WriteLine("   ❌ Field names must exist in your index");
            Console.WriteLine("   ❌ Searching fewer fields may miss relevant content");
            Console.WriteLine("   ❌ Field-specific search can be less flexible");
            Console.WriteLine("   ❌ Some fields might not be searchable");

            Console.WriteLine("\n🔧 Performance Tips:");
            Console.WriteLine("   ✅ Use field selection to reduce result size");
            Console.WriteLine("   ✅ Search high-value fields first (title, description)");
            Console.WriteLine("   ✅ Combine field search with other filters");
            Console.WriteLine("   ✅ Test which fields give best results for your use case");

            // Demonstrate field strategy
            Console.WriteLine("\n🎯 Field Search Strategy Example:");
            var searchOps = new FieldSearch(searchClient);

            var query = "javascript tutorial";

            Console.WriteLine($"\nQuery: '{query}'");
            Console.WriteLine("Strategy: Search from most specific to most general");

            // 1. Title only (most specific)
            var titleResults = await searchOps.SearchSpecificFieldAsync(query, "title", 1);
            var titleCount = titleResults.GetResults().Count();
            Console.WriteLine($"\n1. Title only: {titleCount} results");

            // 2. Title + description (moderate)
            var titleDescResults = await searchOps.SearchMultipleFieldsAsync(query, new[] { "title", "description" }, 1);
            var titleDescCount = titleDescResults.GetResults().Count();
            Console.WriteLine($"2. Title + Description: {titleDescCount} results");

            // 3. All fields (broadest)
            var allResults = await searchClient.SearchAsync<SearchDocument>(query, new SearchOptions { Size = 1 });
            var allCount = allResults.Value.GetResults().Count();
            Console.WriteLine($"3. All fields: {allCount} results");

            Console.WriteLine($"\n💡 Recommendation: Start specific, broaden if needed");
        }
    }
}

File: docs/beginner/module-02-basic-search/code-samples/javascript/05_field_search.js
/**
 * Field-Specific Search - Module 2 JavaScript Examples
 * Searching within specific fields in Azure AI Search using JavaScript SDK
 * 
 * This module demonstrates:
 * - Searching specific fields
 * - Field selection for results
 * - Multi-field searches
 * - Field weighting concepts
 * - When to use field-specific search
 */

const { SearchClient, AzureKeyCredential } = require('@azure/search-documents');

class FieldSearch {
    /**
     * Initialize with a search client
     * @param {SearchClient} searchClient - Azure Search client
     */
    constructor(searchClient) {
        if (!searchClient) {
            throw new Error('SearchClient is required');
        }
        this.searchClient = searchClient;
    }

    /**
     * Search within a specific field only
     * @param {string} query - Search query string
     * @param {string} field - Field name to search in
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async searchSpecificField(query, field, top = 10) {
        try {
            console.log(`Searching field '${field}' for: '${query}'`);

            const searchOptions = {
                searchFields: [field],
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results in field '${field}'`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error searching field '${field}': ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search within multiple specific fields
     * @param {string} query - Search query string
     * @param {Array<string>} fields - List of field names to search in
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results
     */
    async searchMultipleFields(query, fields, top = 10) {
        try {
            console.log(`Searching fields [${fields.join(', ')}] for: '${query}'`);

            const searchOptions = {
                searchFields: fields,
                top: top,
                includeTotalCount: true
            };

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results in specified fields`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error searching multiple fields: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Search and return only selected fields
     * @param {string} query - Search query string
     * @param {Array<string>} selectFields - List of field names to return in results
     * @param {Array<string>} searchFields - List of field names to search in (optional)
     * @param {number} top - Maximum number of results to return
     * @returns {Promise<Object>} Search results with only selected fields
     */
    async searchWithSelectedFields(query, selectFields, searchFields = null, top = 10) {
        try {
            console.log(`Searching for '${query}' with selected fields: [${selectFields.join(', ')}]`);

            const searchOptions = {
                select: selectFields,
                top: top,
                includeTotalCount: true
            };

            if (searchFields) {
                searchOptions.searchFields = searchFields;
            }

            const results = await this.searchClient.search(query, searchOptions);
            
            const resultArray = [];
            for await (const result of results.results) {
                resultArray.push(result);
            }

            console.log(`Found ${resultArray.length} results with selected fields`);
            return {
                results: resultArray,
                totalCount: results.count
            };
        } catch (error) {
            console.error(`Error in search with selected fields: ${error.message}`);
            return { results: [], totalCount: 0 };
        }
    }

    /**
     * Compare search results across different fields
     * @param {string} query - Search query string
     * @param {Array<string>} fields - List of fields to compare
     * @param {number} top - Maximum results per field
     * @returns {Promise<Object>} Object mapping field names to their results
     */
    async compareFieldSearches(query, fields, top = 5) {
        const results = {};

        // Search all fields (default)
        try {
            const allResults = await this.searchClient.search(query, {
                top: top,
                includeTotalCount: true
            });
            
            const resultArray = [];
            for await (const result of allResults.results) {
                resultArray.push(result);
            }

            results['all_fields'] = {
                results: resultArray,
                totalCount: allResults.count
            };
        } catch (error) {
            console.error(`Error in all fields search: ${error.message}`);
            results['all_fields'] = { results: [], totalCount: 0 };
        }

        // Search individual fields
        for (const field of fields) {
            results[field] = await this.searchSpecificField(query, field, top);
        }

        return results;
    }

    /**
     * Display comparison of field-specific searches
     * @param {string} query - Original search query
     * @param {Object} results - Results from different field searches
     */
    static displayFieldComparison(query, results) {
        console.log(`\n🎯 Field-Specific Search Comparison: '${query}'`);
        console.log('='.repeat(70));

        for (const [fieldName, fieldResults] of Object.entries(results)) {
            const displayName = fieldName.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
            
            console.log(`\n${displayName}:`);
            console.log(`   Results found: ${fieldResults.results.length}`);

            if (fieldResults.results.length > 0) {
                console.log('   Top matches:');
                fieldResults.results.slice(0, 3).forEach((result, index) => {
                    const title = result.document.title || 'No title';
                    const score = result.score || 0.0;
                    console.log(`     ${index + 1}. ${title} (Score: ${score.toFixed(3)})`);
                });

                // Show average score
                const avgScore = fieldResults.results.reduce((sum, r) => sum + (r.score || 0), 0) / fieldResults.results.length;
                console.log(`   Average score: ${avgScore.toFixed(3)}`);
            } else {
                console.log('   No matches found');
            }

            console.log('-'.repeat(50));
        }

        // Analysis
        console.log(`\n📊 FIELD ANALYSIS:`);
        for (const [fieldName, fieldResults] of Object.entries(results)) {
            const count = fieldResults.results.length;
            if (count > 0) {
                const maxScore = Math.max(...fieldResults.results.map(r => r.score || 0));
                console.log(`   ${fieldName}: ${count} results (max score: ${maxScore.toFixed(3)})`);
            } else {
                console.log(`   ${fieldName}: No results`);
            }
        }
    }
}

/**
 * Demonstrate field-specific search operations
 */
async function demonstrateFieldSearch() {
    console.log('🎯 Field-Specific Search Demonstration');
    console.log('='.repeat(50));

    try {
        // Initialize search client (replace with your actual service details)
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const searchOps = new FieldSearch(searchClient);

        // Example 1: Compare field searches
        console.log('\n1️⃣ Field Search Comparison');

        const query = 'python';
        const commonFields = ['title', 'content', 'description', 'author'];

        const fieldResults = await searchOps.compareFieldSearches(query, commonFields, 5);
        FieldSearch.displayFieldComparison(query, fieldResults);

        // Example 2: Multi-field search
        console.log('\n' + '='.repeat(70));
        console.log('\n2️⃣ Multi-Field Search Examples');
        console.log('-'.repeat(30));

        const multiFieldExamples = [
            { description: 'Find tutorials by title or description', fields: ['title', 'description'] },
            { description: 'Search content and tags', fields: ['content', 'tags'] },
            { description: 'Author and title search', fields: ['author', 'title'] }
        ];

        for (const example of multiFieldExamples) {
            console.log(`\n📋 ${example.description}`);
            console.log(`   Fields: ${example.fields.join(', ')}`);

            const results = await searchOps.searchMultipleFields('tutorial', example.fields, 3);
            console.log(`   Results: ${results.results.length} found`);

            if (results.results.length > 0) {
                const topResult = results.results[0];
                const title = topResult.document.title || 'No title';
                const score = topResult.score || 0.0;
                console.log(`   Top match: ${title} (Score: ${score.toFixed(3)})`);
            }
        }

        // Example 3: Field selection
        console.log('\n' + '='.repeat(70));
        console.log('\n3️⃣ Field Selection Examples');
        console.log('-'.repeat(30));

        const selectionExamples = [
            { description: 'Basic info only', selectFields: ['id', 'title', 'author'] },
            { description: 'Content preview', selectFields: ['title', 'description', 'url'] },
            { description: 'Metadata only', selectFields: ['id', 'publishedDate', 'category'] }
        ];

        for (const example of selectionExamples) {
            console.log(`\n📋 ${example.description}`);
            console.log(`   Selected fields: ${example.selectFields.join(', ')}`);

            const results = await searchOps.searchWithSelectedFields(
                'programming', 
                example.selectFields, 
                null, 
                2
            );

            console.log(`   Results: ${results.results.length} found`);

            if (results.results.length > 0) {
                const firstResult = results.results[0];
                const availableFields = Object.keys(firstResult.document).filter(k => !k.startsWith('@'));
                console.log(`   Available fields in result: ${availableFields.join(', ')}`);
            }
        }

        console.log('\n✅ Field-specific search demonstration completed!');

    } catch (error) {
        console.error(`❌ Demo failed: ${error.message}`);
        console.log('Make sure your Azure AI Search service is configured correctly.');
    }
}

/**
 * Demonstrate best practices for field-specific search
 */
async function fieldSearchBestPractices() {
    console.log('\n📚 Field-Specific Search Best Practices');
    console.log('='.repeat(50));

    console.log('\n💡 When to Use Field-Specific Search:');
    console.log('\n✅ Search Specific Fields When:');
    console.log('   - You know exactly which field contains relevant information');
    console.log('   - You want to search titles only (more precise)');
    console.log('   - You need to search metadata fields (author, category, etc.)');
    console.log('   - You want to exclude certain fields from search');

    console.log('\n✅ Select Specific Fields When:');
    console.log('   - You only need certain fields in results (faster)');
    console.log('   - You want to reduce network traffic');
    console.log('   - You\'re building a summary or preview');
    console.log('   - You need consistent result structure');

    console.log('\n⚠️ Field Search Considerations:');
    console.log('   ❌ Field names must exist in your index');
    console.log('   ❌ Searching fewer fields may miss relevant content');
    console.log('   ❌ Field-specific search can be less flexible');
    console.log('   ❌ Some fields might not be searchable');

    console.log('\n🔧 Performance Tips:');
    console.log('   ✅ Use field selection to reduce result size');
    console.log('   ✅ Search high-value fields first (title, description)');
    console.log('   ✅ Combine field search with other filters');
    console.log('   ✅ Test which fields give best results for your use case');

    // Demonstrate field strategy
    console.log('\n🎯 Field Search Strategy Example:');
    
    try {
        const searchClient = new SearchClient(
            'https://your-service.search.windows.net',
            'your-index-name',
            new AzureKeyCredential('your-api-key')
        );

        const searchOps = new FieldSearch(searchClient);
        const query = 'javascript tutorial';

        console.log(`\nQuery: '${query}'`);
        console.log('Strategy: Search from most specific to most general');

        // 1. Title only (most specific)
        const titleResults = await searchOps.searchSpecificField(query, 'title', 1);
        console.log(`\n1. Title only: ${titleResults.results.length} results`);

        // 2. Title + description (moderate)
        const titleDescResults = await searchOps.searchMultipleFields(query, ['title', 'description'], 1);
        console.log(`2. Title + Description: ${titleDescResults.results.length} results`);

        // 3. All fields (broadest)
        const allResults = await searchClient.search(query, { top: 1 });
        let allCount = 0;
        for await (const result of allResults.results) {
            allCount++;
        }
        console.log(`3. All fields: ${allCount} results`);

        console.log(`\n💡 Recommendation: Start specific, broaden if needed`);

    } catch (error) {
        console.error(`❌ Field strategy demo failed: ${error.message}`);
    }
}

// Main execution
async function main() {
    try {
        await demonstrateFieldSearch();
        await fieldSearchBestPractices();

        console.log('\n💡 Next Steps:');
        console.log('   - Try searching different fields in your index');
        console.log('   - Experiment with field combinations');
        console.log('   - Check out other examples for more search options');
        console.log('   - Learn about result processing techniques');
    } catch (error) {
        console.error(`❌ Main execution failed: ${error.message}`);
    }
}

// Export for use in other modules
module.exports = {
    FieldSearch,
    demonstrateFieldSearch,
    fieldSearchBestPractices
};

// Run if this file is executed directly
if (require.main === module) {
    main();
}